Plane: Can now command roll from a companion computer.

mday299 opened the pull request:
This PR allows setting roll from a companion computer using the MAVLINK_MSG_ID_SET_ATTITUDE_TARGET message.

Care has been taken to ensure this message is only accepted in GUIDED mode to ensure failsafes will still function in case of a switch to RTL mode (roll control returns to autopilot).  Also, any mode change resets control to the autopilot.

squilter commented:
Copter uses a timeout for this. If no new attitude messages are sent for 3 (?) seconds, copter will loiter. I think something similar might be a good idea here.

magicrub commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -2103,6 +2103,39 @@ void GCS_MAVLINK_Plane::handleMessage(mavlink_message_t* msg)
         plane.DataFlash.remote_log_block_status_msg(chan, msg);
         break;
 
+    // BEGIN GTRI MOD ////////////
```
please remove // BEGIN GTRI MOD ////////////

mday299 replied:
Removed in updated PR.

magicrub commented on ArduPlane/Plane.h:
```diff
@@ -524,6 +524,7 @@ class Plane : public AP_HAL::HAL::Callbacks {
 
         // are we doing loiter mode as a VTOL?
         bool vtol_loiter:1;
+
```
whitespace

mday299 replied:
Removed in updated PR.

magicrub commented on ArduPlane/Plane.h:
```diff
@@ -553,6 +554,9 @@ class Plane : public AP_HAL::HAL::Callbacks {
     bool throttle_suppressed;
+	
+    // guiding roll from an external controller? (e.g., companion computer)
+    bool guiding_roll;
```
I see you attempted to add guided_roll to auto_state, which was a good line of thinking. Lets make a new guided_state struct and put this into it

mday299 replied:
Created guided_state in updated PR.

magicrub commented on ArduPlane/GCS_Mavlink.cpp:
```diff
@@ -2103,6 +2103,39 @@ void GCS_MAVLINK_Plane::handleMessage(mavlink_message_t* msg)
+
+        if (att_target.type_mask & roll_mask) {
+            // Set the flag for external roll to the nav control
+            plane.guiding_roll = true;
```
what if you only get a single packet with this roll type? Seems the rolling flag is set forevever until you switch modes. As @squilter said, this needs a timeout. Something for the guided_state!

mday299 replied:
guided_state now contains a timing mechanism.

magicrub commented on ArduPlane/Attitude.cpp:
```diff
@@ -529,7 +529,12 @@ void Plane::calc_nav_pitch()
+    //  If so, don't updated nav_roll_cd as it is being guided externally.
+    if( ! plane.guided_state.guiding_roll ||
+          AP_HAL::millis() - plane.guided_state.last_guided_ms > 3000) {
+        nav_roll_cd = constrain_int32(nav_controller->nav_roll_cd(), -roll_limit_cd, roll_limit_cd);
```
I don't think it's wise to remove this roll restriction, it's there for the safety of the aircraft and controlled by params. If they need to be bigger in guided then they are probably ok in AUTO too so the param should change. Can you give a use-case for this?

mday299 replied:
Good call, can't believe I missed that.  I'll work that and update the PR.

mday299 commented:
Updated the PR. Now uses a struct vice a bool.  Now times out after 3 seconds of silence from the external controller.  Upon timing out the plane returns to the location where external control began and loiters there.

magicrub commented on ArduPlane/system.cpp:
```diff
@@ -336,6 +336,9 @@ void Plane::set_mode(enum FlightMode mode)
     crash_state.impact_detected = false;
 
+    // reset external attitude guidance
+    guided_state.guiding_roll = false;
```
after thinking about this, we could easily do this with a single variable, the boolean is not needed.

set the timer to 0 here to clear/reset, and when using it just check for non-zero

magicrub replied:
I can change that when merging

mday299 replied:
@magicrub  That would preclude future changes for pitch and yaw.

Chat on gitter?

mday299 commented:
Updated the PR considering @magicrub comments.  External control now respects bank limits.  Also added the externally commanded roll to the guided_state struct.

magicrub commented:
merged via https://github.com/ArduPilot/ardupilot/commit/8ac433f9913b1685a6e16b4b18581362c0773b00 and https://github.com/ArduPilot/ardupilot/commit/85afd9e2455bb8dd8f1288b9d84f8f5b5ec85bb4

Thanks @mday299 !
